src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+            // Reader event is or is about to be signaled
+            public const int ReaderSignaled = 0x400;
+            // Writer event is or is about to be signaled
+            public const int WriterSignaled = 0x800;
+            public const int Writer = 0x1000;
+            // Waiting reader increment
+            public const int WaitingReader = 0x2000;
+            // Max number of waiting readers (maximum count must be less than or equal to the maximum count of readers)
+            public const int WaitingReadersMask = 0x7FE000;
+            public const int WaitingReadersShift = 13;
+            // Waiting writer increment
+            public const int WaitingWriter = 0x800000;
+            // Max number of waiting writers
+            public const int WaitingWritersMask = unchecked((int)0xFF800000);
+            // Events are being cached (for all intents and purposes, "cached" means "disposed of"). New acquire requests cannot
+            // become waiters during this time since they need the events for waiting. Once events are disposed of and the
+            // state is changed, new to-be-waiters can recreate the events they need.
+            public const int CachingEvents = ReaderSignaled | WriterSignaled;
+        }
+
+        // REVIEW: The original code maintained lists of thread-local lock entries on the CLR's thread objects, and manually
+        // released lock entries, which involved walking through all threads. While this is possible with ThreadLocal<T>, I
+        // preferred to use a similar design to that from ReaderWriterLockSlim, and allow reusing empty entries without removing
+        // entries, since it is unlikely that the list length for any thread would get unreasonably long. I have preserved the
+        // logic from the original code to move recently used entries to the front of the list in the GetOrCreate path.
+        /// <summary>
+        /// Stores thread-local lock info and manages the association of this info with each <see cref="ReaderWriterLock"/>
+        /// owned by a thread.
+        /// </summary>
+        private class LockEntry
+        {
+            [ThreadStatic]
+            private static LockEntry t_lockEntryHead;
+
+            private long _lockID;
+            private LockEntry _next;
+            public ushort _readerLevel;
+
+            private LockEntry(long lockID)
+            {
+                _lockID = lockID;
+            }
+
+            public bool HasLockID(long lockID)
+            {
+                return _lockID == lockID;
+            }
+
+            public bool IsEmpty
+            {
+                get
+                {
+                    return _readerLevel == 0;
+                }
+            }
+
+            private static void VerifyNoNonemptyEntryInListAfter(long lockID, LockEntry afterEntry)
+            {
+                Debug.Assert(lockID != 0);
+                Debug.Assert(afterEntry != null);
+
+#if DEBUG
```

Comment:
You can call it, but the call site is removed in release builds. It's exactly how Debug.Assert is implemented to be removed in release.